compiler/next/include/chpl/resolution/resolution-queries.h

```diff
@@ -21,19 +21,142 @@
 
 #include "chpl/resolution/resolution-types.h"
+#include "chpl/resolution/scope-types.h"
 
 namespace chpl {
 namespace resolution {
 
 
-  /*
-  // Resolves the top-level declarations in a module
-  const ResolvedSymbol&
-  resolveModule(Context* context, const uast::Module* mod);
+  ////// resolution basics
 
-  const ResolvedSymbolVec& resolveFile(Context* context, UniqueString path);
+  /**
+    Resolve the contents of a Module
    */
+  const ResolutionResultByPostorderID& resolveModule(Context* context, ID id);
 
+  /**
+    Compute the type for a NamedDecl with a particular id.
+   */
+  const types::QualifiedType& typeForModuleLevelSymbol(Context* context, ID id);
+
+  /**
+    Compute the type for a Builtin type using just its name
+   */
+  const types::QualifiedType& typeForBuiltin(Context* context, UniqueString name);
+
+  /////// function resolution
+
+  /**
+    Compute an UntypedFnSignature for a Function, or return
+    nullptr if the passed ID is not a Function.
+    The pointer result is unique'd and safe to use as a query argument.
+   */
+  const UntypedFnSignature* untypedSignature(Context* context, ID id);
+
+
+  /**
+    Compute a TypedFnSignature from an UntypedFnSignature.
+    The TypedFnSignature will represent generic and potentially unknown
+    types if the function is generic.
+
+    The pointer result is unique'd and safe to use as a query argument.
+   */
+  const TypedFnSignature*
+  typedSignatureInitial(Context* context,
+                        const UntypedFnSignature* untypedSig);
+
+  /**
+    Instantiate a TypedFnSignature from
+     * the result of typedSignatureInitial,
+     * a CallInfo describing the types at the call site, and
+     * a point-of-instantiation scope representing the POI scope of the call
+
+    Returns nullptr if the instantiation failed.
+   */
+  const TypedFnSignature* instantiateSignature(Context* context,
+                                               const TypedFnSignature* sig,
+                                               CallInfo call,
+                                               const PoiScope* poiScope);
+
+  /**
+    Compute a ResolvedFunction given a TypedFnSignature.
+    Checks the generic cache for potential for reuse. When reuse occurs,
+    the ResolvedFunction might point to a different TypedFnSignature.
+   */
+  const ResolvedFunction* resolveFunction(Context* context,
+                                          const TypedFnSignature* sig,
+                                          const PoiScope* poiScope);
+
+
+  /**
+    Resolves a concrete function using the above queries.
+    */
+  const ResolvedFunction* resolveConcreteFunction(Context* context, ID id);
+
+  /**
+    Returns the ResolvedFunction called by a particular
+    ResolvedExpression, if there was exactly one candidate.
+    This function does not handle return intent overloading.
+   */
+  const ResolvedFunction* resolveOnlyCandidate(Context* context,
```

Comment:
What happens if ResolvedExpr has more than one candidate? Does this assert, throw an exception, or return a nullptr?

Reply:
nullptr. I adjusted the comment.